torchtune/datasets/_chat.py

```diff
@@ -0,0 +1,181 @@
+        )
+
+        assert len(prompt_tokens) == len(label_tokens)
+
+        return prompt_tokens, label_tokens
+
+    def _get_turns(
+        self, dialogue: Dialogue
+    ) -> Generator[Tuple[Dict[str, str], str], None, None]:
+        prompt_messages = {}
+        for message in dialogue:
+            # If we are at the assistant message, we are at the end of a turn, yield.
+            if message["role"] == "assistant":
+                if "user" not in prompt_messages:
+                    raise ValueError(
+                        f"Missing a user message before assistant message: {message['content']}"
+                    )
+                yield prompt_messages, message["content"]
+                prompt_messages = {}
+            # Otherwise, continue to add to the turn's messages
+            else:
+                if message["role"] in prompt_messages:
+                    raise ValueError(
+                        f"Duplicate {message['role']} message in dialogue: {message['content']}"
+                    )
+                prompt_messages[message["role"]] = message["content"]
+
+        # If we never yielded, then the last turn was incomplete
+        if prompt_messages:
+            raise ValueError(
+                f"Incomplete turn in dialogue, current turn: {prompt_messages}"
+            )
+
+
+def chat_dataset(
+    tokenizer: Tokenizer,
+    source: str,
+    conversation_format: str,
+    template: str,
+    max_seq_len: int,
+    train_on_input: bool = False,
+    **load_dataset_kwargs: Dict[str, Any],
+) -> ChatDataset:
+    """
+    Build a configurable dataset with conversations. This method should be
+    used to configure a custom chat dataset from the yaml config instead of
+    using `ChatDataset` directly, as it is made to be config friendly.
+
+    Args:
+        tokenizer (Tokenizer): Tokenizer used to encode data. Tokenize must implement an `encode` and `decode` method.
+        source (str): path string of dataset, anything supported by HuggingFace's `load_dataset`
+            (https://huggingface.co/docs/datasets/en/package_reference/loading_methods#datasets.load_dataset.path)
+        conversation_format (str): string specifying expected format of conversations in the dataset
+            for automatic conversion to the llama format. Supported formats are: "sharegpt"
+        template (str): class name of template used to format the prompt.
+        max_seq_len (int): Maximum number of tokens in the returned input and label token id lists.
+        train_on_input (bool): Whether the model is trained on the prompt or not. Default is False.
+        **load_dataset_kwargs (Dict[str, Any]): additional keyword arguments to pass to `load_dataset`.
+
+    Returns:
+        ChatDataset: the configured ChatDataset
+
+    Raises:
+        ValueError: if the conversation format is not supported
+    """
+    if conversation_format == "sharegpt":
+        convert_to_dialogue = sharegpt_to_llama2_dialogue
```

Comment:
Is this imported anywhere?

Reply:
huh, all my tests ran successfully on locally. maybe i need to push